timelock-impl/src/main/java/com/palantir/atlasdb/backup/AtlasBackupResource.java

```diff
@@ -117,6 +121,45 @@ private InProgressBackupToken prepareBackup(Namespace namespace) {
                 .build();
     }
 
+    @Override
+    public ListenableFuture<RefreshBackupResponse> refreshBackup(AuthHeader authHeader, RefreshBackupRequest request) {
+        return handleExceptions(() -> refreshBackupInternal(authHeader, request));
+    }
+
+    private ListenableFuture<RefreshBackupResponse> refreshBackupInternal(
+            AuthHeader authHeader, RefreshBackupRequest request) {
+        if (!authHeaderValidator.suppliedTokenIsValid(authHeader)) {
+            log.error(
+                    "Attempted to complete backup with an invalid auth header. "
+                            + "The provided token must match the configured permitted-backup-token.",
+                    SafeArg.of("request", request));
+            throw new ServiceException(ErrorType.PERMISSION_DENIED);
+        }
+
+        Map<InProgressBackupToken, ListenableFuture<Optional<RefreshLockResponseV2>>> futureMap =
+                request.getTokens().stream().collect(Collectors.toMap(token -> token, this::refreshBackupAsync));
+        ListenableFuture<Map<InProgressBackupToken, RefreshLockResponseV2>> singleFuture =
+                AtlasFutures.allAsMap(futureMap, MoreExecutors.newDirectExecutorService());
+
+        return Futures.transform(singleFuture, this::getRefreshedTokens, MoreExecutors.directExecutor());
+    }
```

Comment:
Good question. Happily, the direct executor runs the tasks in the same thread as the one that's dealing with the future, so we don't have additional threads here.

The javadoc (which is worth reading btw) does caution against chaining many of these tasks together as it could lead to StackOverflow; we have a couple of transforms, but not enough to cause any issues.

Reply:
(noting that we should check timelock thread usage during backup scale tests)

Reply:
Thanks, that makes a lot of sense.